pkg/validator/promql_expression.go

```diff
@@ -167,6 +168,73 @@ func (h expressionUsesOnlyAllowedLabelsForMetricRegexp) Validate(_ unmarshaler.R
 	return errs
 }
 
+// newExpressionUsesOnlyAllowedLabelValuesForMetricRegexp.
+type expressionUsesOnlyAllowedLabelValuesForMetricRegexp struct {
+	allowedLabelValues map[string][]string
+	metricNameRegexp   *regexp.Regexp
+}
+
+func newExpressionUsesOnlyAllowedLabelValuesForMetricRegexp(paramsConfig yaml.Node) (Validator, error) {
+	params := struct {
+		allowedLabelValues map[string][]string `yaml:"allowedLabelValues"`
+		MetricNameRegexp   string              `yaml:"metricNameRegexp"`
+	}{}
+	if err := paramsConfig.Decode(&params); err != nil {
+		return nil, err
+	}
+	if len(params.allowedLabelValues) == 0 {
+		return nil, fmt.Errorf("missing allowed label values")
+	}
+	compiled, err := compileAnchoredRegexp(params.MetricNameRegexp)
+	if err != nil {
+		return nil, fmt.Errorf("invalid metric name regexp %s: %w", params.MetricNameRegexp, err)
+	}
+	return &expressionUsesOnlyAllowedLabelValuesForMetricRegexp{allowedLabelValues: params.allowedLabelValues, metricNameRegexp: compiled}, nil
+}
+
+func (h expressionUsesOnlyAllowedLabelValuesForMetricRegexp) String() string {
+	var allowedLabelValues string
+	for label, values := range h.allowedLabelValues {
+		allowedLabelValues += fmt.Sprintf(" %s: [%s]", label, strings.Join(values, ","))
+	}
+	return fmt.Sprintf("for metrics matching regexp '%s', given lables use only specified values:%s\n", h.metricNameRegexp, allowedLabelValues)
+}
+
+func (h expressionUsesOnlyAllowedLabelValuesForMetricRegexp) Validate(_ unmarshaler.RuleGroup, rule rulefmt.Rule, _ *prometheus.Client) []error {
+	matchers, err := getLabelMatchersForMetricRegexp(rule.Expr, h.metricNameRegexp)
+	errs := []error{}
+	if err != nil {
+		errs = append(errs, err)
+		return errs
+	}
+	for _, matcher := range matchers {
+		var allowedLabels []string
+		var ok, matchFound bool
+		if allowedLabels, ok = h.allowedLabelValues[matcher.Name]; !ok {
+			// no whitelist is configured for this label
+			continue
+		}
+		switch matcher.Type {
+		// It does not make sense to validate label values whitelist for negative matchers
+		case labels.MatchNotEqual:
+			continue
+		case labels.MatchNotRegexp:
+			continue
+		default:
```

Comment:
nit: the default could be dropped and move its code after the switch to reduce nesting

Reply:
reafactored